src/jit/decomposelongs.cpp

```diff
@@ -0,0 +1,1206 @@
+// DecompNodeHelper: fgWalkTreePost callback helper for LONG decomposition
+//
+// Arguments:
+//    ppTree - tree node we are working on.
+//    data - tree walk context, with data->pCallbackData as a DecomposeLongs*
+//
+// Return Value:
+//    Standard tree walk result.
+//
+// static
+Compiler::fgWalkResult DecomposeLongs::DecompNodeHelper(GenTree** ppTree, Compiler::fgWalkData* data)
+{
+    DecomposeLongs* decomp = (DecomposeLongs*)data->pCallbackData;
+    decomp->DecomposeNode(ppTree, data);
+    return Compiler::WALK_CONTINUE;
+}
+
+
+//------------------------------------------------------------------------
+// DecomposeNode: Decompose long-type trees into lower and upper halves.
+//
+// Arguments:
+//    *ppTree - A node that may or may not require decomposition.
+//    data    - The tree-walk data that provides the context.
+//
+// Return Value:
+//    None. It the tree at *ppTree is of TYP_LONG, it will generally be replaced.
+//
+void DecomposeLongs::DecomposeNode(GenTree** ppTree, Compiler::fgWalkData* data)
+{
+    GenTree* tree = *ppTree;
+
+    // Handle the case where we are implicitly using the lower half of a long lclVar.
+    if ((tree->TypeGet() == TYP_INT) && tree->OperIsLocal())
+    {
+        LclVarDsc* varDsc = m_compiler->lvaTable + tree->AsLclVarCommon()->gtLclNum;
+        if (varTypeIsLong(varDsc) && varDsc->lvPromoted)
+        {
+#ifdef DEBUG
+            if (m_compiler->verbose)
+            {
+                printf("Changing implicit reference to lo half of long lclVar to an explicit reference of its promoted half:\n");
+                m_compiler->gtDispTree(tree);
+            }
+#endif // DEBUG
+            m_compiler->lvaDecRefCnts(tree);
+            unsigned loVarNum = varDsc->lvFieldLclStart;
+            tree->AsLclVarCommon()->SetLclNum(loVarNum);
+            m_compiler->lvaIncRefCnts(tree);
+            return;
+        }
+    }
+
+    if (tree->TypeGet() != TYP_LONG)
+    {
+        return;
+    }
+
+#ifdef DEBUG
+    if (m_compiler->verbose)
+    {
+        printf("Decomposing TYP_LONG tree.  BEFORE:\n");
+        m_compiler->gtDispTree(tree);
+    }
+#endif // DEBUG
+
+    switch (tree->OperGet())
+    {
+    case GT_PHI:
+    case GT_PHI_ARG:
+        break;
+
+    case GT_LCL_VAR:
+        DecomposeLclVar(ppTree, data);
+        break;
+
+    case GT_LCL_FLD:
+        DecomposeLclFld(ppTree, data);
+        break;
+
+    case GT_STORE_LCL_VAR:
+        DecomposeStoreLclVar(ppTree, data);
+        break;
+
+    case GT_CAST:
+        DecomposeCast(ppTree, data);
+        break;
+
+    case GT_CNS_LNG:
+        DecomposeCnsLng(ppTree, data);
+        break;
+
+    case GT_CALL:
+        DecomposeCall(ppTree, data);
+        break;
+
+    case GT_RETURN:
+        assert(tree->gtOp.gtOp1->OperGet() == GT_LONG);
+        break;
+
+    case GT_STOREIND:
+        DecomposeStoreInd(ppTree, data);
+        break;
+
+    case GT_STORE_LCL_FLD:
```

Comment:
Can we put these in their own functions too, so that when we fill them in, we just fill out the function? (here, GT_IND, for sure. Not sure on the arithmetic operations, because those may be morphed into helper calls in morph rather than handled in Decomp, like some div/mul operations already are).